napari/_qt/qt_main_window.py

```diff
@@ -949,29 +981,26 @@ def _rebuild_plugins_menu(self, event=None):
 
         # Add a menu item (QAction) for each available plugin widget
         for hook_type, (plugin_name, widgets) in plugin_manager.iter_widgets():
-            multiprovider = len(widgets) > 1
-            if multiprovider:
-                menu = QMenu(plugin_name, self._qt_window)
-                self.plugins_menu.addMenu(menu)
-            else:
-                menu = self.plugins_menu
+            self._add_plugin_menu_items(hook_type, plugin_name, widgets)
 
-            for wdg_name in widgets:
-                key = (plugin_name, wdg_name)
-                if multiprovider:
-                    action = QAction(wdg_name, parent=self._qt_window)
-                else:
-                    full_name = plugin_menu_item_template.format(*key)
-                    action = QAction(full_name, parent=self._qt_window)
-
-                def _add_widget(*args, key=key, hook_type=hook_type):
-                    if hook_type == 'dock':
-                        self.add_plugin_dock_widget(*key)
-                    else:
-                        self._add_plugin_function_widget(*key)
+    def _add_plugin_menu_items(self, hook_type, plugin_name, widgets):
+        multiprovider = len(widgets) > 1
+        if multiprovider:
+            menu = QMenu(plugin_name, self._qt_window)
+            self.plugins_menu.addMenu(menu)
+        else:
+            menu = self.plugins_menu
 
-                menu.addAction(action)
-                action.triggered.connect(_add_widget)
+        for wdg_name in widgets:
+            key = (plugin_name, wdg_name)
+            if hook_type == 'dock':
+                self.add_plugin_dock_widget(
```

Comment:
yeah, here's the problem.  Rather than adding an action that calls the local `_add_widget` function from before, you're now immediately instantiating (and just hiding) *every* widget in the environment.  Definitely can't have that.

`add_plugin_dock_widget` is a public method, and shouldn't be used to populate menus... I also don't like the leaking of the internal logic (menu=None, multiprovider=False) into the signature of that method

Reply:
thanks, @tlambert03 I will take a closer look soon.  About the `menu` and `multiprovider`, I was trying to have a way to easily include submenus of toggleActions on the plugins menu.  I will think a bit more about how to do this a better way.